### DISCUSSION TITLE
Random choice from set with probabilities

### DISCUSSION THREAD

**Discussion by buddha314** [mathnet/mathnet-numerics#840] [Ideas]
Very simple idea implemented in most languages, but apparently not in C# or MathNet yet.

I have a set `{A, B, C}` and I want to pick an element from it with probabilities `{0.1, 0.75, 0.15}`.  Python uses and argument to `random.choice(x, p=[0.1, 0.75, 0.15])` and that works nicely.

**Comment by PasaOpasen** [comment]
Try this:

        public static T GetRandomElementFromArrayWithProbabilities<T>(T[] vals, double[] probs)
        {
            var vers = new double[probs.Length];
            double sum = probs.Sum();

            vers[0] = probs[0] / sum;
            for (int i = 1; i < vers.Length - 1; i++)
            {
                vers[i] = probs[i] / sum + vers[i - 1];
            }
            vers[vers.Length - 1] = 1.0;

            double rndval = randomgen.NextDouble();
            for (int i = 0; i < vers.Length; i++)
                if (vers[i] >= rndval)
                    return vals[i];
            return vals.Last();
        }
